### PR TITLE
Scenes: Show transformations when editing scene dashboard

### PR DESCRIPTION
**Pull request by oscarkilhed**
**What is this feature?**
image

Shows transformations that are applied to the panel data for scene dashboards.
Shows a message and call to action to add transformations when there are no transformations.

This is the start of migrating the transformations tab. Transformations can be removed and edited, but it's not possible to add or reorder transformations. Transformations changes will not be saved.

**Special notes for your reviewer:**

The bottom panel containing the query, transformations and alert tabs does not behave like in the old panel editing UX. This means that the bottom panel is not scrollable and takes up a lot of space. Will address this in a different PR.

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: public/app/features/dashboard-scene/panel-edit/PanelDataPane/PanelDataTransformations/PanelDataTransformationsTab.tsx [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+  const dataState = props.sceneDataTransformer.useState();
+  const transformationEditorRows: TransformationsEditorTransformation[] = [];
+
+  let i = 0; // Ids need to be unique for drag to change order to work
+  for (const t of dataState.transformations) {
+    if ('id' in t) {
+      transformationEditorRows.push({ id: `${i} - ${t.id}`, transformation: t });
+      i++;
+    }
+  }
```
Comment (torkelo) [inline public/app/features/dashboard-scene/panel-edit/PanelDataPane/PanelDataTransformations/PanelDataTransformationsTab.tsx:59]:
```suggestion
  const transformations = dataState.transformations.map(( transformation, index) => ({ id: `${index}`,  transformation})  
```

Comment (oscarkilhed) [reply public/app/features/dashboard-scene/panel-edit/PanelDataPane/PanelDataTransformations/PanelDataTransformationsTab.tsx:59]: I wrote it this way to have the type guard discern between `DataTransformerConfig` and `CustomTransformOperator`. From my understanding when editing dashboard scenes we would never really see `CustomTransformOperator` in the transformations array. I could explore other ways to get around this type issue.

File: public/app/features/dashboard-scene/panel-edit/PanelDataPane/PanelDataTransformations/PanelDataTransformationsTab.tsx [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    return 'Transformations';
+  }
+
+  getItemsCount() {
+    const dataProvider = this._panelManager.state.panel.state.$data;
+    if (dataProvider instanceof SceneDataTransformer) {
+      return dataProvider.state.transformations.length;
+    } else {
+      return null;
+    }
```
Comment (dprokop) [inline public/app/features/dashboard-scene/panel-edit/PanelDataPane/PanelDataTransformations/PanelDataTransformationsTab.tsx:37]: Think we can do sth similar to what we did for queries tab

```ts
getItemsCount() {
  return this.getQueries().length;
}
```

The getQueries method uses queryRunner getter under the hood that will throw if there's no query runner for a panel. That saves us checks for the object existance or whether or not it's of a correct type.

File: public/app/features/dashboard-scene/panel-edit/VizPanelManager.tsx [hunk 3]
```diff
@@ -217,6 +218,14 @@ export class VizPanelManager extends SceneObjectBase<VizPanelManagerState> {
     runner.setState({ queries });
   }
 
+  public changeTransformations(transformations: DataTransformerConfig[]) {
+    const dataProvider = this.state.panel.state.$data;
+    if (dataProvider instanceof SceneDataTransformer) {
+      dataProvider.setState({ transformations });
+      dataProvider.reprocessTransformations();
+    }
```
Comment (dprokop) [inline public/app/features/dashboard-scene/panel-edit/VizPanelManager.tsx:226]: Hmm, think that would be nice to be handled in the SceneDataTransformer state change handler? WDYT @torkelo ?

**Comment by torkelo** [comment]
pushed some proposed refactorings to https://github.com/grafana/grafana/compare/oscark/show-transformations-when-editing-scene-dashboard...scene-transform-refactors?expand=1  to try to simplify it a bit

**Review by torkelo — COMMENTED** [review]
I would skip the PanelDataTransformations sub folder the folder structure is a bit too nested already (dashboard-scene/panel-edit/PanelDataPanel/PanelDataTransformations)  

PanelDataPane is not that full of files yet :)

File: public/app/features/dashboard-scene/panel-edit/VizPanelManager.tsx (hunk 3 above)
Comment (torkelo) [reply public/app/features/dashboard-scene/panel-edit/VizPanelManager.tsx:226]: yea, maybe but not sure it's needed. It's not that common yet to update transformations, and having this manual call feels ok, it's similar with SceneQueryRunner in that you have to call runQueries after updating queries

File: public/app/features/dashboard-scene/panel-edit/PanelDataPane/PanelDataTransformations/PanelDataTransformationsTab.tsx [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+  if (!(panelState.$data instanceof SceneDataTransformer)) {
+    return;
+  }
+
+  const dataState = panelState.$data.useState();
+
+  return (
+    <CustomScrollbar autoHeightMin="100%">
+      <Container>
```
Comment (torkelo) [inline public/app/features/dashboard-scene/panel-edit/PanelDataPane/PanelDataTransformations/PanelDataTransformationsTab.tsx:102]: can we move these CustomScrollbar + Container to PanelDataPaneRendered (so shared between all tabs). 

Then we can simplify do:

```ts
if (dataState.transformations.length < 1) {
  return <EmptyTransformationsMessage onShowPicker={() => {}} />
}
```

I really struggle reading react markup with long multiline co ternary operators :)

**Comment by oscarkilhed** [comment]
@torkelo @dprokop Thank you, I've updated the PR with the feedback

**Review by torkelo — APPROVED** [review]
Looks good!

File: public/app/features/dashboard-scene/panel-edit/PanelDataPane/PanelDataTransformationsTab.tsx [hunk 5]
```diff
@@ -32,15 +38,104 @@ export class PanelDataTransformationsTab
...
-  // const dataObj = dataRef.resolve();
-  // // const { transformations } = dataObj.useState();
+interface TransformationEditorProps {
+  transformations: DataTransformerConfig[];
+  model: PanelDataTransformationsTab;
+  data: PanelData;
+}
 
-  return <div>TODO Transformations</div>;
+function TransformationsEditor({ transformations, model, data }: TransformationEditorProps) {
```
[the remark this answers is not in the document]
Comment (oscarkilhed) [reply public/app/features/dashboard-scene/panel-edit/PanelDataPane/PanelDataTransformationsTab.tsx]: Makes sense

**Pull request merged**